server/src/main/java/io/crate/execution/dml/upsert/TransportShardUpsertAction.java

```diff
@@ -386,9 +386,12 @@ protected WriteReplicaResult processRequestItemsOnReplica(IndexShard indexShard,
                 SequenceNumbers.UNASSIGNED_PRIMARY_TERM
             );
             IndexResult result = indexShard.index(index);
-            assert result.getResultType() != Engine.Result.Type.MAPPING_UPDATE_REQUIRED
-                : "If parsedDoc.newColumns is empty there must be no mapping update requirement";
-            location = result.getTranslogLocation();
+            if (result.getResultType() != Type.SUCCESS) {
+                assert false : "doc-level index failure must not happen on replica";
+                throw Exceptions.toRuntimeException(result.getFailure());
```

Comment:
Maybe logging this error is helpful as well.

Reply:
It also gets logged in https://github.com/crate/crate/blob/e14384978f22cb8b1f9edff311e004db479ee6d1/server/src/main/java/org/elasticsearch/index/shard/IndexShard.java#L874-L885

Reply:
shall we turn this into an error or at least warn?

Reply:
I think that makes things more confusing because for the primary operation the errors can be expected (e.g. version conflict, which then triggers retry handling, etc..)

With the change the client would now also receive the error, so we have logs + client reporting. I don't think more is necessary. Especially considering that the change for the replica is merely a santiy check.